// @layr/component-http-server/dist/node-cjs/component-http-server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComponentHTTPServer = void 0;
const tslib_1 = require("tslib");
const koa_1 = tslib_1.__importDefault(require("koa"));
const koa_logger_1 = tslib_1.__importDefault(require("koa-logger"));
const koa_json_error_1 = tslib_1.__importDefault(require("koa-json-error"));
const cors_1 = tslib_1.__importDefault(require("@koa/cors"));
const component_server_1 = require("@layr/component-server");
const component_koa_middleware_1 = require("@layr/component-koa-middleware");
const debug_1 = tslib_1.__importDefault(require("debug"));
const debug = debug_1.default('layr:component-http-server');
// To display the debug log, set this environment:
// DEBUG=layr:component-http-server DEBUG_DEPTH=10
const DEFAULT_PORT = 3333;
/**
 * A class allowing to serve a root [`Component`](https://layrjs.com/docs/v1/reference/component) so it can be accessed by a [`ComponentHTTPClient`](https://layrjs.com/docs/v1/reference/component-http-client).
 *
 * This class provides a basic HTTP server providing one endpoint to serve your root component. If you wish to build an HTTP server providing multiple endpoints, you can use a middleware such as [`component-express-middleware`](https://layrjs.com/docs/v1/reference/component-express-middleware), or implement the necessary plumbing to integrate a [`ComponentServer`](https://layrjs.com/docs/v1/reference/component-server) in your custom HTTP server.
 *
 * #### Usage
 *
 * Create an instance of `ComponentHTTPServer` by specifying the root [`Component`](https://layrjs.com/docs/v1/reference/component) you want to serve, and use the [`start()`](https://layrjs.com/docs/v1/reference/component-http-server#start-instance-method) method to start the server.
 *
 * See an example of use in [`ComponentHTTPClient`](https://layrjs.com/docs/v1/reference/component-http-client).
 */
class ComponentHTTPServer {
    /**
     * Creates a component HTTP server.
     *
     * @param componentOrComponentServer The root [`Component`](https://layrjs.com/docs/v1/reference/component) class to serve. An instance of a [`ComponentServer`](https://layrjs.com/docs/v1/reference/component-server) will be created under the hood. Alternatively, you can pass an existing instance of a [`ComponentServer`](https://layrjs.com/docs/v1/reference/component-server).
     * @param [options.port] A number specifying the TCP port to listen to (default: `3333`).
     * @param [options.version] A number specifying the version of the created [`ComponentServer`](https://layrjs.com/docs/v1/reference/component-server) (default: `undefined`).
     *
     * @returns A `ComponentHTTPServer` instance.
     *
     * @category Creation
     */
    constructor(componentOrComponentServer, options = {}) {
        const componentServer = component_server_1.ensureComponentServer(componentOrComponentServer, options);
        const { port = DEFAULT_PORT, limit, delay, errorRate } = options;
        this._componentServer = componentServer;
        this._port = port;
        this._serveComponentOptions = { limit, delay, errorRate };
    }
    /**
     * Starts the component HTTP server.
     *
     * @example
     * ```
     * const server = new ComponentHTTPServer(Movie, {port: 3210});
     *
     * await server.start();
     * ```
     *
     * @category Methods
     * @async
     */
    start() {
        if (this._server !== undefined) {
            throw new Error('The component HTTP server is already started');
        }
        const koa = new koa_1.default();
        koa.use(koa_logger_1.default((message) => {
            debug(message);
        }));
        koa.use(koa_json_error_1.default());
        koa.use(cors_1.default({ maxAge: 900 })); // 15 minutes
        koa.use(component_koa_middleware_1.serveComponent(this._componentServer, this._serveComponentOptions));
        return new Promise((resolve) => {
            this._server = koa.listen(this._port, () => {
                debug(`Component HTTP server started (port: ${this._port})`);
                resolve();
            });
        });
    }
    /**
     * Stops the component HTTP server.
     *
     * @category Methods
     * @async
     */
    stop() {
        const server = this._server;
        if (server === undefined) {
            throw new Error('The component HTTP server is not started');
        }
        return new Promise((resolve) => {
            server.close(() => {
                this._server = undefined;
                debug(`Component HTTP server stopped`);
                resolve();
            });
        });
    }
}
exports.ComponentHTTPServer = ComponentHTTPServer;
