// simple-cloning/dist/node-esm/cloning.js
import isObjectLike from 'lodash/isObjectLike';
import mapValues from 'lodash/mapValues';
export function clone(value, options) {
    if (Array.isArray(value)) {
        return cloneArray(value, options);
    }
    if (isObjectLike(value)) {
        return cloneObject(value, options);
    }
    if (typeof value === 'function') {
        throw new Error('Cannot clone a function');
    }
    return value;
}
function cloneObject(object, options) {
    const objectCloner = options === null || options === void 0 ? void 0 : options.objectCloner;
    if (objectCloner !== undefined) {
        const clonedObject = objectCloner(object);
        if (clonedObject !== undefined) {
            return clonedObject;
        }
    }
    if (object instanceof Date) {
        return cloneDate(object);
    }
    if (object instanceof RegExp) {
        return cloneRegExp(object);
    }
    if (object instanceof Error) {
        return cloneError(object, options);
    }
    return cloneAttributes(object, options);
}
function cloneArray(array, options) {
    return array.map((item) => clone(item, options));
}
function cloneDate(date) {
    return new Date(date);
}
function cloneRegExp(regExp) {
    return new RegExp(regExp.source, regExp.flags);
}
function cloneError(error, options) {
    const clonedError = new Error(error.message);
    Object.assign(clonedError, cloneAttributes(error, options));
    return clonedError;
}
function cloneAttributes(object, options) {
    return mapValues(object, (value) => clone(value, options));
}
